File: QrCodeScannerWindows/Tests/ArbiterTests.cs
using System.Reflection;
using QrScanner;
using QrScanner.Tests;

// Checks the rule this app exists for: which of two printed QR codes is taken.
//
// No test framework, because this needs none: it runs as part of the build and
// fails it on the first wrong answer. The checks are the macOS suite's, in the
// same order, so the two scanners cannot drift apart on the one decision that
// matters.

internal static class ArbiterTests
{
    private static int _failures;
    private static int _checks;

    private const string Factor =
        "A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4"
        + "A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4"
        + "A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4"
        + "A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4A1B2C3D4";

    private static int Main()
    {
        BothCodesReadableIsAcceptedImmediately();
        OneDamagedCodeFallsBackToTheReadableOne();
        ALoneCodeIsNotAcceptedOnItsFirstSighting();
        DisagreeingCodesAreRefused();
        ADuplicateReportIsNotASecondCode();
        InvalidGeometryCannotCorroborate();
        FlickerDoesNotResetTheTally();
        AChangedPayloadRestartsTheTally();
        PayloadInspection();
        BothLanguagesAreComplete();
        DecodesRealCodes();

        if (_failures == 0)
        {
            Console.WriteLine($"ArbiterTests: {_checks} checks passed");
            return 0;
        }

        Console.Error.WriteLine($"ArbiterTests: {_failures} of {_checks} checks failed");
        return 1;
    }

    private static void Expect(bool condition, string description)
    {
        _checks++;
        if (!condition)
        {
            _failures++;
            Console.Error.WriteLine($"FAIL {description}");
        }
    }

    private static void ExpectEqual<T>(T actual, T expected, string description)
    {
        _checks++;
        if (!EqualityComparer<T>.Default.Equals(actual, expected))
        {
            _failures++;
            Console.Error.WriteLine($"FAIL {description}: expected {expected}, got {actual}");
        }
    }

    private static void ExpectThrows<TException>(Action action, string description)
        where TException : Exception
    {
        try
        {
            action();
            Expect(false, description);
        }
        catch (TException)
        {
            Expect(true, description);
        }
    }

    /// <summary>Two codes side by side, as they are printed on a sheet.</summary>
    private static Detection[] Pair(string left, string right) =>
    [
        new Detection(left, 0.3, 0.5),
        new Detection(right, 0.7, 0.5),
    ];

    private static Detection[] Single(string payload) => [new Detection(payload, 0.5, 0.5)];

    private static void ExpectAccepted(ScanOutcome outcome, string payload, CorroborationKind kind, int count, string description)
    {
        Expect(
            outcome.Kind == ScanOutcomeKind.Accepted
            && string.Equals(outcome.Payload, payload, StringComparison.Ordinal)
            && outcome.Corroboration.Kind == kind
            && outcome.Corroboration.Count == count,
            $"{description} (got {outcome.Kind}/{outcome.Corroboration.Kind}/{outcome.Corroboration.Count})");
    }

    private static void ExpectConfirming(ScanOutcome outcome, int progress, int required, string description)
    {
        Expect(
            outcome.Kind == ScanOutcomeKind.Confirming
            && outcome.Progress == progress
            && outcome.Required == required,
            $"{description} (got {outcome.Kind} {outcome.Progress}/{outcome.Required})");
    }

    /// <summary>
    /// Both printed codes decode. They agree, so the payload is taken at once -
    /// no repeat count needed, because a second printed copy is better evidence
    /// than the same code read twice.
    /// </summary>
    private static void BothCodesReadableIsAcceptedImmediately()
    {
        var arbiter = new CodeArbiter();
        ScanOutcome outcome = arbiter.Admit(Pair(Factor, Factor));
        ExpectAccepted(outcome, Factor, CorroborationKind.MatchingCodes, 2, "two matching codes are accepted on the first frame");
    }

    /// <summary>
    /// The user's case: one code is creased or stained, so it never decodes and
    /// never reaches the arbiter. The readable one carries the content.
    /// </summary>
    private static void OneDamagedCodeFallsBackToTheReadableOne()
    {
        var arbiter = new CodeArbiter(requiredRepeats: 3);
        ExpectConfirming(arbiter.Admit(Single(Factor)), 1, 3, "first read of a lone code confirms once");
        ExpectConfirming(arbiter.Admit(Single(Factor)), 2, 3, "second read of a lone code confirms twice");
        ExpectAccepted(
            arbiter.Admit(Single(Factor)),
            Factor,
            CorroborationKind.RepeatedReads,
            3,
            "a lone code is accepted once it has repeated itself");
    }

    private static void ALoneCodeIsNotAcceptedOnItsFirstSighting()
    {
        var arbiter = new CodeArbiter(requiredRepeats: 8);
        for (int step = 1; step < 8; step++)
        {
            ExpectConfirming(arbiter.Admit(Single(Factor)), step, 8, $"a lone code is still unconfirmed after {step} frames");
        }
    }

    /// <summary>
    /// Two codes that decode to different content cannot both be the sheet's
    /// factor. Nothing is accepted and the disagreement is reported.
    /// </summary>
    private static void DisagreeingCodesAreRefused()
    {
        var arbiter = new CodeArbiter();
        ScanOutcome outcome = arbiter.Admit(Pair("ALPHA", "BETA"));
        Expect(
            outcome.Kind == ScanOutcomeKind.Conflict
            && outcome.ConflictingPayloads is ["ALPHA", "BETA"],
            "disagreeing codes are refused");

        // The conflict must also clear whatever was being counted, so a payload
        // half-way to acceptance cannot be completed after a second code has
        // contradicted it.
        var counting = new CodeArbiter(requiredRepeats: 2);
        _ = counting.Admit(Single("ALPHA"));
        _ = counting.Admit(Pair("ALPHA", "BETA"));
        ExpectConfirming(counting.Admit(Single("ALPHA")), 1, 2, "a conflict restarts the tally");
    }

    /// <summary>
    /// The same physical code reported twice in one frame must not be mistaken
    /// for two printed copies; otherwise the strongest acceptance the app
    /// offers would rest on a duplicate report.
    /// </summary>
    private static void ADuplicateReportIsNotASecondCode()
    {
        var arbiter = new CodeArbiter(requiredRepeats: 4);
        Detection[] overlapping =
        [
            new Detection(Factor, 0.500, 0.500),
            new Detection(Factor, 0.505, 0.502),
        ];
        ExpectConfirming(arbiter.Admit(overlapping), 1, 4, "two reports of one code are not treated as two codes");
    }

    private static void InvalidGeometryCannotCorroborate()
    {
        var arbiter = new CodeArbiter(requiredRepeats: 4);
        Detection[] oneRealOneInvalid =
        [
            new Detection(Factor, 0.5, 0.5),
            new Detection(Factor, double.NaN, 0.8),
        ];
        ExpectConfirming(
            arbiter.Admit(oneRealOneInvalid),
            1,
            4,
            "a NaN decoder position cannot masquerade as a second printed code");

        var conflictingInvalid = new CodeArbiter();
        ExpectEqual(
            conflictingInvalid.Admit(
                [new Detection(Factor, 0.5, 0.5), new Detection("different", double.NaN, 0.8)]).Kind,
            ScanOutcomeKind.Conflict,
            "malformed geometry cannot hide a conflicting decoded payload");

        var invalidOnly = new CodeArbiter();
        ExpectEqual(
            invalidOnly.Admit([new Detection(Factor, double.PositiveInfinity, 0.5)]).Kind,
            ScanOutcomeKind.Searching,
            "a detection without finite in-frame geometry is ignored");
        ExpectThrows<ArgumentOutOfRangeException>(
            () => _ = new CodeArbiter(minimumSeparation: double.NaN),
            "a NaN separation threshold is rejected");
        ExpectThrows<ArgumentOutOfRangeException>(
            () => _ = new CodeArbiter(minimumSeparation: double.PositiveInfinity),
            "an infinite separation threshold is rejected");
    }

    /// <summary>
    /// Detection drops out between frames. A single missing frame must not
    /// throw away a tally, or a lone code would never reach the repeat count.
    /// </summary>
    private static void FlickerDoesNotResetTheTally()
    {
        var arbiter = new CodeArbiter(requiredRepeats: 3, toleratedMisses: 2);
        _ = arbiter.Admit(Single(Factor));
        _ = arbiter.Admit([]);
        _ = arbiter.Admit(Single(Factor));
        ExpectAccepted(
            arbiter.Admit(Single(Factor)),
            Factor,
            CorroborationKind.RepeatedReads,
            3,
            "a dropped frame does not discard the tally");

        var patient = new CodeArbiter(requiredRepeats: 3, toleratedMisses: 2);
        _ = patient.Admit(Single(Factor));
        _ = patient.Admit([]);
        _ = patient.Admit([]);
        _ = patient.Admit([]);
        ExpectConfirming(patient.Admit(Single(Factor)), 1, 3, "a code out of view long enough starts over");
    }

    private static void AChangedPayloadRestartsTheTally()
    {
        var arbiter = new CodeArbiter(requiredRepeats: 3);
        _ = arbiter.Admit(Single("ALPHA"));
        _ = arbiter.Admit(Single("ALPHA"));
        ExpectConfirming(arbiter.Admit(Single("BETA")), 1, 3, "a different payload starts its own tally");
    }

    private static void PayloadInspection()
    {
        Expect(PayloadInspector.Accept("") is null, "an empty payload is rejected");
        Expect(
            PayloadInspector.Accept(new string('x', PayloadInspector.MaximumLength + 1)) is null,
            "an oversized payload is rejected");
        ExpectEqual(PayloadInspector.Accept(Factor), Factor, "a normal payload passes through unchanged");

        // Nothing is trimmed: the copy button must hand over exactly what the
        // code contained, so padding is reported rather than removed.
        ExpectEqual(PayloadInspector.Accept("  spaced  "), "  spaced  ", "whitespace is preserved, not trimmed");
        Expect(
            PayloadInspector.Notices("  spaced  ").Any(notice => notice.Kind == PayloadNoticeKind.SurroundingWhitespace),
            "surrounding whitespace is reported");
        Expect(
            PayloadInspector.Notices("a‮b").Any(notice =>
                notice.Kind == PayloadNoticeKind.BidirectionalOverrides && notice.Count == 1),
            "a direction override is reported");
        Expect(
            PayloadInspector.Notices("ab").Any(notice =>
                notice.Kind == PayloadNoticeKind.ControlCharacters && notice.Count == 1),
            "a control character is reported");
        Expect(PayloadInspector.Notices(Factor).Count == 0, "a plain hexadecimal factor raises nothing");
        ExpectEqual(
            PayloadInspector.DisplayText("ab"),
            "a‹U+0007›b",
            "control characters are made visible in the display copy");
        Expect(
            PayloadInspector.Notices("a​b").Any(notice =>
                notice.Kind == PayloadNoticeKind.InvisibleCharacters && notice.Count == 1),
            "a zero-width character is reported");
        string unpairedSurrogate = "a\uD800b";
        Expect(
            PayloadInspector.Notices(unpairedSurrogate).Any(notice =>
                notice.Kind == PayloadNoticeKind.ControlCharacters && notice.Count == 1),
            "an unpaired UTF-16 surrogate is reported instead of crashing inspection");
        ExpectEqual(
            PayloadInspector.DisplayText(unpairedSurrogate),
            "a‹U+D800›b",
            "an unpaired UTF-16 surrogate is rendered visibly");
    }

    /// <summary>
    /// Walks both language tables and checks that nothing was left blank or
    /// left in the other language.
    /// </summary>
    /// <remarks>
    /// The plain strings are reached through reflection rather than listed
    /// here, so a string added to Strings tomorrow is covered without this test
    /// being touched - which is the only way a completeness check stays true.
    /// </remarks>
    private static void BothLanguagesAreComplete()
    {
        PropertyInfo[] properties = typeof(Strings)
            .GetProperties(BindingFlags.Public | BindingFlags.Instance)
            .Where(property => property.PropertyType == typeof(string))
            .ToArray();

        Expect(properties.Length > 0, "the string table exposes plain strings");

        foreach (PropertyInfo property in properties)
        {
            var german = (string?)property.GetValue(Strings.German);
            var english = (string?)property.GetValue(Strings.English);
            Expect(!string.IsNullOrWhiteSpace(german), $"German {property.Name} is not blank");
            Expect(!string.IsNullOrWhiteSpace(english), $"English {property.Name} is not blank");

            // Nothing but the app's own name should be identical in both
            // tables; a match anywhere else is a translation never written.
            if (property.Name is not ("WindowTitle" or "CopyButton"))
            {
                Expect(
                    !string.Equals(german, english, StringComparison.Ordinal),
                    $"{property.Name} is the same in both languages - untranslated");
            }
        }

        // The delegates cannot be reached the same way, so they are called.
        (string Label, Func<Strings, string> Render)[] cases =
        [
            ("Confirming", strings => strings.Confirming(3, 8)),
            ("Conflict", strings => strings.Conflict(2)),
            ("AcceptedMatching", strings => strings.AcceptedMatching(2)),
            ("AcceptedRepeated", strings => strings.AcceptedRepeated(8)),
            ("Copied", strings => strings.Copied(90)),
            ("AccessDenied", strings => strings.AccessDenied("a")),
            ("NoticeControlCharacters", strings => strings.NoticeControlCharacters(1)),
            ("NoticeBidirectional", strings => strings.NoticeBidirectional(1)),
            ("NoticeInvisible", strings => strings.NoticeInvisible(1)),
        ];

        foreach ((string label, Func<Strings, string> render) in cases)
        {
            foreach (Language language in Enum.GetValues<Language>())
            {
                Expect(render(Strings.For(language)).Length > 0, $"{label} renders in {language}");
            }
        }

        // The number has to survive into the sentence in both languages, which
        // is the failure a hand-written translation actually makes.
        Expect(Strings.German.Copied(90).Contains("90", StringComparison.Ordinal), "the German clipboard notice states the delay");
        Expect(Strings.English.Copied(90).Contains("90", StringComparison.Ordinal), "the English clipboard notice states the delay");
        Expect(Strings.German.Confirming(3, 8).Contains('3', StringComparison.Ordinal), "the German progress notice states the count");
        Expect(Strings.English.Confirming(3, 8).Contains('8', StringComparison.Ordinal), "the English progress notice states the target");

        ExpectEqual(Enum.GetValues<Language>().Length, 2, "two languages are offered");
        foreach (Language language in Enum.GetValues<Language>())
        {
            Expect(Strings.OwnName(language).Length > 0, $"{language} names itself for the switch");
        }

        // Every outcome has a sentence; a missing one would show the ready text
        // where a conflict should be reported.
        foreach (Language language in Enum.GetValues<Language>())
        {
            Strings strings = Strings.For(language);
            Expect(
                strings.Describe(ScanOutcome.Conflict(["A", "B"])).Contains('2', StringComparison.Ordinal),
                $"{language} reports how many codes disagreed");
            Expect(
                strings.Describe(ScanOutcome.Accepted("X", CorroborationKind.MatchingCodes, 2))
                    != strings.Describe(ScanOutcome.Accepted("X", CorroborationKind.RepeatedReads, 8)),
                $"{language} distinguishes two matching codes from one repeated code");
        }
    }

    /// <summary>
    /// Runs the decoder over a rendered sheet, the way the camera would.
    /// </summary>
    /// <remarks>
    /// Everything above tests the rule in the abstract. This puts two real QR
    /// codes into a real image, decodes them through the same call the frame
    /// handler makes, and checks that the arbiter is handed two separate
    /// detections rather than one - which is the difference between the app's
    /// strongest acceptance and its weakest.
    /// </remarks>
    private static void DecodesRealCodes()
    {
        const int size = 320;
        const int gap = 40;
        int width = (2 * size) + (3 * gap);
        int height = size + (2 * gap);
        byte[] frame = QrImage.RenderPair(Factor, Factor, size, gap, width, height);

        var session = new ScanSession(new SynchronizationContext());
        IReadOnlyList<Detection> detections = session.Decode(frame, width, height);

        Expect(detections.Count == 2, $"both printed codes are decoded from one frame (got {detections.Count})");
        Expect(
            detections.All(detection => string.Equals(detection.Payload, Factor, StringComparison.Ordinal)),
            "both decoded codes carry the printed factor");
        if (detections.Count == 2)
        {
            Expect(
                Math.Abs(detections[0].CenterX - detections[1].CenterX) > 0.02,
                "the two codes are reported at separate positions");
            ExpectAccepted(
                new CodeArbiter().Admit(detections),
                Factor,
                CorroborationKind.MatchingCodes,
                2,
                "a real sheet is accepted as confirmed by two codes");
        }

        // One code destroyed: the decoder returns nothing for it rather than
        // returning damaged content, so the readable one is left on its own.
        byte[] damaged = QrImage.RenderPair(Factor, Factor, size, gap, width, height, destroyLeft: true);
        IReadOnlyList<Detection> single = session.Decode(damaged, width, height);
        Expect(single.Count == 1, $"a destroyed code drops out instead of decoding wrongly (got {single.Count})");
        if (single.Count == 1)
        {
            Expect(
                string.Equals(single[0].Payload, Factor, StringComparison.Ordinal),
                "the surviving code still carries the whole factor");
        }
    }
}

File: QrCodeScanner/Tests/ArbiterTests.swift
import AppKit
import CoreGraphics
import Foundation

// Checks the rule this app exists for: which of two printed QR codes is taken.
//
// No test framework, because the build has no package manager and this needs
// none: it runs as part of the build and fails it on the first wrong answer.

private var failures = 0
private var checks = 0

private func expect(
    _ condition: Bool,
    _ description: String,
    file: StaticString = #file,
    line: UInt = #line
) {
    checks += 1
    if !condition {
        failures += 1
        FileHandle.standardError.write(Data("FAIL \(description) (line \(line))\n".utf8))
    }
}

private func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ description: String,
    line: UInt = #line
) {
    checks += 1
    if actual != expected {
        failures += 1
        FileHandle.standardError.write(
            Data("FAIL \(description): expected \(expected), got \(actual) (line \(line))\n".utf8))
    }
}

/// Two codes side by side, as they are printed on a sheet.
private func pair(_ left: String, _ right: String) -> [Detection] {
    [
        Detection(payload: left, center: CGPoint(x: 0.3, y: 0.5)),
        Detection(payload: right, center: CGPoint(x: 0.7, y: 0.5)),
    ]
}

private func single(_ payload: String) -> [Detection] {
    [Detection(payload: payload, center: CGPoint(x: 0.5, y: 0.5))]
}

private let factor = String(repeating: "A1B2C3D4", count: 16)

@main
@MainActor
private enum ArbiterTests {
    static func main() {
        bothCodesReadableIsAcceptedImmediately()
        oneDamagedCodeFallsBackToTheReadableOne()
        aLoneCodeIsNotAcceptedOnItsFirstSighting()
        disagreeingCodesAreRefused()
        aDuplicateReportIsNotASecondCode()
        flickerDoesNotResetTheTally()
        aChangedPayloadRestartsTheTally()
        payloadInspection()
        scanSessionLifecycle()
        singleInstancePolicy()
        clipboardLifecycle()
        bothLanguagesAreComplete()

        if failures == 0 {
            print("ArbiterTests: \(checks) checks passed")
            exit(0)
        }
        FileHandle.standardError.write(Data("ArbiterTests: \(failures) of \(checks) checks failed\n".utf8))
        exit(1)
    }

    static func scanSessionLifecycle() {
        var lifecycle = ScanSessionLifecycle()
        expect(!lifecycle.acceptsDetections, "a new scan session rejects frames before camera start")
        lifecycle.resume()
        expect(lifecycle.acceptsDetections, "a resumed scan session accepts detections")
        lifecycle.suspend()
        expect(!lifecycle.acceptsDetections, "an accepted payload suspends further detections")
        lifecycle.resume()
        expect(lifecycle.acceptsDetections, "a cleared result starts a fresh scan")
        lifecycle.stop()
        expect(!lifecycle.acceptsDetections, "closing the window stops detection delivery")
        expect(!lifecycle.canResume, "closing the window makes the scan session terminal")
        expect(!lifecycle.resume(), "a delayed camera-start callback cannot resume a stopped session")
        expect(!lifecycle.acceptsDetections, "resume after stop cannot re-enable detection delivery")
        lifecycle.suspend()
        expect(!lifecycle.acceptsDetections, "suspend after stop cannot weaken the terminal state")

        let terminationGate = ScanSessionTerminationGate()
        expect(!terminationGate.isStopped, "a new off-main camera gate permits initial start")
        terminationGate.stop()
        expect(terminationGate.isStopped, "the off-main camera gate remains terminal after stop")
    }

    static func clipboardLifecycle() {
        let name = NSPasteboard.Name("de.michael-feinermann.qr-scanner.tests.\(UUID().uuidString)")
        let pasteboard = NSPasteboard(name: name)
        pasteboard.clearContents()
        defer {
            pasteboard.clearContents()
            pasteboard.releaseGlobally()
        }

        var expiryCallbacks = 0
        let clipboard = VolatileClipboard(lifetime: 3600, pasteboard: pasteboard)
        clipboard.copy("factor-secret") { expiryCallbacks += 1 }
        expectEqual(
            pasteboard.string(forType: .string),
            "factor-secret",
            "the requested payload reaches the isolated test pasteboard")
        expect(
            pasteboard.types?.contains(NSPasteboard.PasteboardType("org.nspasteboard.ConcealedType")) == true,
            "the copied payload carries the concealed clipboard marker")
        clipboard.clearIfStillOurs()
        expect(pasteboard.string(forType: .string) == nil, "the scanner clears a clipboard item it still owns")
        expectEqual(expiryCallbacks, 1, "clearing the owned payload reports one expiry")

        clipboard.copy("second-secret") { expiryCallbacks += 1 }
        pasteboard.clearContents()
        pasteboard.setString("user-copy", forType: .string)
        clipboard.clearIfStillOurs()
        expectEqual(
            pasteboard.string(forType: .string),
            "user-copy",
            "expiry never erases a clipboard value copied later by the user")
        expectEqual(expiryCallbacks, 2, "losing clipboard ownership still closes the scanner lifecycle once")
    }

    static func singleInstancePolicy() {
        expect(
            SingleInstancePolicy.shouldContinue(currentPID: 100, runningPIDs: []),
            "the first scanner process owns the camera/payload instance")
        expect(
            SingleInstancePolicy.shouldContinue(currentPID: 100, runningPIDs: [100, 101]),
            "the oldest scanner PID remains the single-instance winner")
        expect(
            !SingleInstancePolicy.shouldContinue(currentPID: 101, runningPIDs: [100, 101]),
            "a direct second executable instance is rejected")
        expect(
            !SingleInstancePolicy.shouldContinue(currentPID: 102, runningPIDs: [100, 101]),
            "multiple observed peers cannot make a later scanner instance win")

        let root = FileManager.default.temporaryDirectory
            .appendingPathComponent("qr-scanner-lock-\(UUID().uuidString)", isDirectory: true)
        do {
            try FileManager.default.createDirectory(at: root, withIntermediateDirectories: false)
            let path = root.appendingPathComponent("instance.lock").path
            var first = try SingleInstancePolicy.acquireLock(at: path)
            expect(first != nil, "the first process atomically acquires the scanner instance lock")
            let second = try SingleInstancePolicy.acquireLock(at: path)
            expect(second == nil, "a simultaneous second process is refused by the kernel lock")
            first = nil
            let replacement = try SingleInstancePolicy.acquireLock(at: path)
            expect(replacement != nil, "the scanner instance lock is released with its process lease")
        } catch {
            expect(false, "the scanner instance-lock regression failed: \(error)")
        }
        try? FileManager.default.removeItem(at: root)
    }

    /// Both printed codes decode. They agree, so the payload is taken at once —
    /// no repeat count needed, because a second printed copy is better evidence
    /// than the same code read twice.
    static func bothCodesReadableIsAcceptedImmediately() {
        var arbiter = CodeArbiter()
        let outcome = arbiter.admit(pair(factor, factor))
        expectEqual(
            outcome,
            .accepted(payload: factor, corroboration: .matchingCodes(count: 2)),
            "two matching codes are accepted on the first frame")
    }

    /// The user's case: one code is creased or stained, so it never decodes and
    /// never reaches the arbiter. The readable one carries the content.
    static func oneDamagedCodeFallsBackToTheReadableOne() {
        var arbiter = CodeArbiter(requiredRepeats: 3)
        var outcome = arbiter.admit(single(factor))
        expectEqual(outcome, .confirming(progress: 1, required: 3), "first read of a lone code confirms once")
        outcome = arbiter.admit(single(factor))
        expectEqual(outcome, .confirming(progress: 2, required: 3), "second read of a lone code confirms twice")
        outcome = arbiter.admit(single(factor))
        expectEqual(
            outcome,
            .accepted(payload: factor, corroboration: .repeatedReads(count: 3)),
            "a lone code is accepted once it has repeated itself")
    }

    static func aLoneCodeIsNotAcceptedOnItsFirstSighting() {
        var arbiter = CodeArbiter(requiredRepeats: 8)
        for step in 1 ..< 8 {
            let outcome = arbiter.admit(single(factor))
            expectEqual(
                outcome,
                .confirming(progress: step, required: 8),
                "a lone code is still unconfirmed after \(step) frames")
        }
    }

    /// Two codes that decode to different content cannot both be the sheet's
    /// factor. Nothing is accepted and the disagreement is reported.
    static func disagreeingCodesAreRefused() {
        var arbiter = CodeArbiter()
        let outcome = arbiter.admit(pair("ALPHA", "BETA"))
        expectEqual(outcome, .conflict(payloads: ["ALPHA", "BETA"]), "disagreeing codes are refused")

        // The conflict must also clear whatever was being counted, so a payload
        // half-way to acceptance cannot be completed after a second code has
        // contradicted it.
        var counting = CodeArbiter(requiredRepeats: 2)
        _ = counting.admit(single("ALPHA"))
        _ = counting.admit(pair("ALPHA", "BETA"))
        let after = counting.admit(single("ALPHA"))
        expectEqual(after, .confirming(progress: 1, required: 2), "a conflict restarts the tally")
    }

    /// The same physical code reported twice in one frame must not be mistaken
    /// for two printed copies; otherwise the strongest acceptance the app
    /// offers would rest on a duplicate report.
    static func aDuplicateReportIsNotASecondCode() {
        var arbiter = CodeArbiter(requiredRepeats: 4)
        let overlapping = [
            Detection(payload: factor, center: CGPoint(x: 0.500, y: 0.500)),
            Detection(payload: factor, center: CGPoint(x: 0.505, y: 0.502)),
        ]
        let outcome = arbiter.admit(overlapping)
        expectEqual(
            outcome,
            .confirming(progress: 1, required: 4),
            "two reports of one code are not treated as two codes")
    }

    /// Detection drops out between frames. A single missing frame must not
    /// throw away a tally, or a lone code would never reach the repeat count.
    static func flickerDoesNotResetTheTally() {
        var arbiter = CodeArbiter(requiredRepeats: 3, toleratedMisses: 2)
        _ = arbiter.admit(single(factor))
        _ = arbiter.admit([])
        _ = arbiter.admit(single(factor))
        let outcome = arbiter.admit(single(factor))
        expectEqual(
            outcome,
            .accepted(payload: factor, corroboration: .repeatedReads(count: 3)),
            "a dropped frame does not discard the tally")

        var patient = CodeArbiter(requiredRepeats: 3, toleratedMisses: 2)
        _ = patient.admit(single(factor))
        _ = patient.admit([])
        _ = patient.admit([])
        _ = patient.admit([])
        let restarted = patient.admit(single(factor))
        expectEqual(
            restarted,
            .confirming(progress: 1, required: 3),
            "a code out of view long enough starts over")
    }

    static func aChangedPayloadRestartsTheTally() {
        var arbiter = CodeArbiter(requiredRepeats: 3)
        _ = arbiter.admit(single("ALPHA"))
        _ = arbiter.admit(single("ALPHA"))
        let outcome = arbiter.admit(single("BETA"))
        expectEqual(
            outcome,
            .confirming(progress: 1, required: 3),
            "a different payload starts its own tally")
    }

    static func payloadInspection() {
        expect(PayloadInspector.accept("") == nil, "an empty payload is rejected")
        expect(
            PayloadInspector.accept(String(repeating: "x", count: PayloadInspector.maximumLength + 1)) == nil,
            "an oversized payload is rejected")
        expectEqual(PayloadInspector.accept(factor), factor, "a normal payload passes through unchanged")

        // Nothing is trimmed: the copy button must hand over exactly what the
        // code contained, so padding is reported rather than removed.
        expectEqual(PayloadInspector.accept("  spaced  "), "  spaced  ", "whitespace is preserved, not trimmed")
        expect(
            PayloadInspector.notices(for: "  spaced  ").contains(.surroundingWhitespace),
            "surrounding whitespace is reported")
        expect(
            PayloadInspector.notices(for: "a\u{202E}b").contains(.bidirectionalOverrides(count: 1)),
            "a direction override is reported")
        expect(
            PayloadInspector.notices(for: "a\u{0007}b").contains(.controlCharacters(count: 1)),
            "a control character is reported")
        expect(
            PayloadInspector.notices(for: factor).isEmpty,
            "a plain hexadecimal factor raises nothing")
        expectEqual(
            PayloadInspector.displayText(for: "a\u{0007}b"),
            "a‹U+0007›b",
            "control characters are made visible in the display copy")
    }

    /// Walks both language tables and checks that nothing was left blank or
    /// left in the other language.
    ///
    /// The plain strings are reached through Mirror rather than listed here, so
    /// a string added to Strings tomorrow is covered without this test being
    /// touched — which is the only way a completeness check stays true.
    static func bothLanguagesAreComplete() {
        let german = Mirror(reflecting: Strings.german)
        let english = Mirror(reflecting: Strings.english)

        var germanPlain: [String: String] = [:]
        var englishPlain: [String: String] = [:]
        for child in german.children {
            if let label = child.label, let value = child.value as? String {
                germanPlain[label] = value
            }
        }
        for child in english.children {
            if let label = child.label, let value = child.value as? String {
                englishPlain[label] = value
            }
        }

        expect(!germanPlain.isEmpty, "the German table exposes plain strings")
        expectEqual(germanPlain.count, englishPlain.count, "both tables carry the same plain strings")

        for (label, value) in germanPlain {
            expect(!value.trimmingCharacters(in: .whitespaces).isEmpty, "German \(label) is not blank")
            guard let counterpart = englishPlain[label] else {
                expect(false, "English is missing \(label)")
                continue
            }
            expect(!counterpart.trimmingCharacters(in: .whitespaces).isEmpty, "English \(label) is not blank")
        }

        // Nothing but the app's own name should be identical in both tables; a
        // match anywhere else is a translation that was never written.
        for (label, value) in germanPlain where label != "menuCopy" {
            if let counterpart = englishPlain[label], counterpart == value {
                expect(false, "\(label) is the same in both languages — untranslated")
            }
        }

        // The closures cannot be reached through Mirror, so they are called.
        let cases: [(String, (Strings) -> String)] = [
            ("confirming", { $0.confirming(3, 8) }),
            ("conflict", { $0.conflict(2) }),
            ("acceptedMatching", { $0.acceptedMatching(2) }),
            ("acceptedRepeated", { $0.acceptedRepeated(8) }),
            ("copied", { $0.copied(90) }),
            ("accessDenied", { $0.accessDenied("a", "b") }),
            ("noticeControlCharacters", { $0.noticeControlCharacters(1) }),
            ("noticeBidirectional", { $0.noticeBidirectional(1) }),
            ("noticeInvisible", { $0.noticeInvisible(1) }),
        ]
        for (label, render) in cases {
            for language in Language.allCases {
                let rendered = render(language.strings)
                expect(!rendered.isEmpty, "\(label) renders in \(language.rawValue)")
            }
        }

        // The number has to survive into the sentence in both languages, which
        // is the failure a hand-written translation actually makes.
        expect(Strings.german.copied(90).contains("90"), "the German clipboard notice states the delay")
        expect(Strings.english.copied(90).contains("90"), "the English clipboard notice states the delay")
        expect(Strings.german.confirming(3, 8).contains("3"), "the German progress notice states the count")
        expect(Strings.english.confirming(3, 8).contains("8"), "the English progress notice states the target")

        expectEqual(Language.allCases.count, 2, "two languages are offered")
        for language in Language.allCases {
            expect(!language.ownName.isEmpty, "\(language.rawValue) names itself for the switch")
        }
    }
}
